docs/codegen-tour.md

```diff
@@ -0,0 +1,843 @@
+# Code generation in oxidize
+
+This document is an attempt to describe in reasonable detail the general
+architecture of the [`read-fonts`][] and [`write-fonts`][] crates, focusing
+specifically on parts that are auto-generated.
+
+> ***note***:
+>
+> at various points in this document I will make use of blockquotes (like this
+one) to highlight
+> particular aspects of the design that may be interesting, confusing, or
+require refinement.
+
+## contents
+
+- [overview](#overview)
+- [`read-fonts`](#read-fonts)
+    - [scalars and `BigEndian<T>`](#scalars-detour)
+    - [tables](#read-tables)
+        - [`FontRead` and `FontReadWithArgs`](#font-read-args)
+        - [versioned tables](#versioned-tables)
+        - [multi-format tables](#multi-format-tables)
+        - [getters](#table-getters)
+        - [offset getters](#offset-getters)
+        - [offset data](#offset-data)
+    - [records](#records)
+        - [zerocopy](#zerocopy)
+        - [copy-on-read](#copy-on-read)
+        - [offsets in records](#offsets-in-records)
+    - [flags and enums](#flags-and-enums)
+    - [traversal](#traversal)
+- [`write-fonts`](#write-fonts)
+    - [tables and records](#write-tables-records)
+    - [fields and `#[compile(..)]`](#table-fields)
+    - [offsets](#write-offsets)
+    - [parsing and `FromTableRef`](#write-parsing)
+    - [validation](#validation)
+    - [compilation and `FontWrite`](#compilation)
+
+## <a id="overview"></a> overview
+
+These two crates can be thought of as siblings, and they both follow the same
+basic high-level design pattern: they contain a set of generated types, mapping
+*as closely as possible* to the types in the [OpenType spec][opentype],
+alongside hand-written code that uses and is used by those types.
+
+The [`read-fonts`][] crate is focused on efficient read access and parsing, and
+the [`write-fonts`][] crate is focused on compilation. The two crates contain a
+parallel `tables` module, with a nearly identical set of type definitions: for
+instance, [both crates][read-name-record] [contain a][write-name-record] `tables::name::NameRecord` type.
+
+We will examine each of these crates separately.
+
+## <a id="read-fonts"></a> `read-fonts`
+
+In the [`read-fonts`][] crate, we make a distinction between *table* objects and
+*record* objects, and we generate different code for each.
```

Comment:
It would be helpful to explain what the difference is, and if they are *always* different. Is it *only* offset resolution or is there more to it?

If a table and a record have identical fields and none of them are offsets to sub-structures would they be identical?

Reply:
I hope this is explained in the subsequent sections, where I walk through each case separately, let me know if that is unclear :)

Reply:
Is it necessary to make me work for it? - as reader I would have appreciated a terse explanation right away.